### PR TITLE
Add round check to consensus gate, fix ordering gate round

### PR DESCRIPTION
**Pull request by lebdron**
### Description of the Change
- Ordering gate was generating incorrect reject rounds based on initial stored value. Now it uses the value from observable
- Yac gate was accepting votes and consensus results for old rounds. Now it ignores them

### Benefits
Predictable pipeline behavior

### Possible Drawbacks 
None?

### Usage Examples or Tests *[optional]*
YacGateTest.OlderVote, YacGateTest.OlderCommit, YacGateTest.OlderReject

### CODE REVIEW

File: irohad/ordering/impl/on_demand_common.cpp [hunk 1]
```diff
@@ -21,5 +21,13 @@ namespace iroha {
     const consensus::RejectRoundType kNextCommitRoundConsumer =
         kFirstRejectRound;
 
+    consensus::Round nextCommitRound(consensus::Round round) {
```
Comment (muratovv) [inline irohad/ordering/impl/on_demand_common.cpp]: Maybe rework it with const ref?

File: irohad/ordering/impl/on_demand_ordering_gate.hpp [hunk 2]
```diff
@@ -43,7 +43,9 @@ namespace iroha {
       /**
        * Represents no storage modification. Reject round increment
        */
-      struct EmptyEvent {};
+      struct EmptyEvent {
+        consensus::Round round;
```
Comment (muratovv) [inline irohad/ordering/impl/on_demand_ordering_gate.hpp]: Add meaning of field here.

File: test/module/irohad/consensus/yac/yac_gate_test.cpp [hunk 3]
```diff
@@ -253,3 +247,104 @@ TEST_F(YacGateTest, DifferentCommit) {
 
   ASSERT_TRUE(gate_wrapper.validate());
 }
+
+/**
+ * @given yac gate with current round initialized
+ * @when vote for older round is called
+ * @then vote is ignored
+ */
+TEST_F(YacGateTest, OlderVote) {
```
Comment (muratovv) [inline test/module/irohad/consensus/yac/yac_gate_test.cpp]: Seems that this and further tests have the same structure. I think better to move initialization logic to a new fixture based on `YacGateTest`.

File: irohad/consensus/yac/impl/yac_gate_impl.cpp [hunk 4]
```diff
@@ -33,12 +33,24 @@ namespace iroha {
...
             consensus_result_cache_(std::move(consensus_result_cache)),
-            log_(logger::log("YacGate")) {
+            log_(logger::log("YacGate")),
+            current_hash_() {
         block_creator_->onBlock().subscribe(
             [this](const auto &event) { this->vote(event); });
       }
 
       void YacGateImpl::vote(const simulator::BlockCreatorEvent &event) {
+        if (not(current_hash_.vote_round < event.round)) {
```
[the remark this answers is not in the document]
Comment (lebdron) [reply irohad/consensus/yac/impl/yac_gate_impl.cpp]: Found a solution with `boost::less_than_comparable`

**Pull request merged**